### PR TITLE
feat(sharepoint): pass through ACL permission metadata

### PR DESCRIPTION
**Pull request by brndnblck**
## Summary
- Extract SharePoint permission data from the Graph API and normalize to the standard `[{read: {users, groups}}, {update: ...}, {delete: ...}]` schema used by Google Drive and Confluence connectors.
- Fetch permissions via Graph JSON batching (`/$batch`, up to 20 per round-trip) with automatic per-item fallback on batch failure.
- Parse identities from raw `permission.properties` JSON to avoid cross-field data bleed in the office365 typed accessors.

## Details
- Added `MICROSOFT_ROLE_MAPPING` covering Graph standard roles and SharePoint-specific `sp.*` roles.
- Added `_extract_identity_ids_from_raw()` to pull Azure AD user/group IDs from `grantedToV2` (preferred) with `grantedTo` fallback. SharePoint `siteGroup` numeric IDs are excluded (not resolvable via Graph).
- Added `extract_permissions()` to normalize raw Graph permissions into the canonical schema.
- Added `_fetch_permissions_batched()` with queue-drain guard on batch failure and `ClientRequestException` handling per item in fallback.
- Refactored `run_async` to window-and-yield in chunks of 20 instead of materializing the full file list.
- 32 unit tests covering identity extraction, role mapping, batching, fallback, and edge cases.

## Limitations
- Requires SharePoint Online (M365) with Azure AD OAuth. SharePoint Server on-prem is not supported (no Graph API).
- Graph API does not support `$expand=permissions` on DriveItem collections, so permissions are fetched separately per batch.
- ACL metadata is a best-effort snapshot; it is not runtime authorization.

Closes PLU-331

---

> [!NOTE]
> **Medium Risk**
> Touches SharePoint indexing flow and adds external Graph batching/fallback logic; failures could impact metadata completeness or indexing performance, though behavior is largely additive and covered by extensive tests.
> 
> **Overview**
> Adds **SharePoint ACL pass-through** by fetching Graph `DriveItem` permissions and normalizing them into the canonical `[{read},{update},{delete}]` `permissions_data` schema.
> 
> Implements Graph `/$batch` permission hydration (20 items per batch) with a per-item fallback on batch failure, plus raw-JSON identity extraction and a `MICROSOFT_ROLE_MAPPING` to translate Microsoft/SharePoint roles into operations. Updates `run_async` to process files in permission-hydrated chunks, and extends unit + integration fixtures to validate deterministic, merged ACL output.
> 
> Bumps package version to `1.5.0` and documents the enhancement in `CHANGELOG.md`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7cb9a084aca246b3f21af01296853a566356087d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: unstructured_ingest/processes/connectors/sharepoint.py [hunk 1]
```diff
@@ -98,6 +125,135 @@ class SharepointIndexer(OnedriveIndexer):
...
+            return [{}]
+
+        normalized: dict[str, dict[str, set[str]]] = {
+            "read": {"users": set(), "groups": set()},
+            "update": {"users": set(), "groups": set()},
+            "delete": {"users": set(), "groups": set()},
+        }
+
+        for perm in permissions:
+            raw_props = perm.properties
```
Comment (potter-potter) [inline unstructured_ingest/processes/connectors/sharepoint.py]: I think you want `perm.to_json()` here

perm.properties returns office365's internal _properties dict where nested fields like grantedToV2 are stored as typed ClientValue objects (e.g. SharePointIdentitySet), not raw dicts — so identity_set.get("user") in _extract_identity_ids_from_raw blows up with AttributeError: 'SharePointIdentitySet' object has no attribute 'get'. perm.to_json() recursively serializes those typed objects back to plain dicts/lists, which both matches the docstring's "raw Graph API permission JSON" assumption and lets the existing .get(...) accessors work unchanged.

File: unstructured_ingest/processes/connectors/sharepoint.py [hunk 2]
```diff
@@ -98,6 +125,135 @@ class SharepointIndexer(OnedriveIndexer):
...
+    ) -> None:
+        """Batch-fetch permissions for a list of drive items using Graph /$batch.
+
+        Queues a GET for each item's permissions collection, then flushes them
+        in a single HTTP call. On batch failure, falls back to per-item fetches
+        so a single 403 does not lose permissions for the entire batch.
+        """
+        from office365.runtime.client_request_exception import ClientRequestException
+
+        for drive_item in drive_items:
```
Comment (potter-potter) [inline unstructured_ingest/processes/connectors/sharepoint.py:239]: drive_item.permissions is properties.get("permissions", PermissionCollection(...)) The default isn't stored back, so each access returns a fresh empty collection. The one we call .get() on here is ephemeral and discarded after execute_batch, so when drive_item_to_file_data_sync later does list(drive_item.permissions) it reads from a brand-new empty collection and permissions_data stays None.

File: unstructured_ingest/processes/connectors/sharepoint.py (hunk 1 above)
Comment (brndnblck) [reply unstructured_ingest/processes/connectors/sharepoint.py]: Addressed in 786d01834c7d9ab8d85f302847551cd0a3c18c51

File: unstructured_ingest/processes/connectors/sharepoint.py (hunk 2 above)
Comment (brndnblck) [reply unstructured_ingest/processes/connectors/sharepoint.py:239]: Addressed in 786d01834c7d9ab8d85f302847551cd0a3c18c51

**Comment by brndnblck** [comment]
Validated end-to-end locally against a live SharePoint tenant using `download_only=True` to isolate the source connector (thank you @potter-potter for the guidance). `permissions_data` is populated on `FileData.metadata` with the expected schema:

```json
"permissions_data": [
  {"read":    {"users": ["e3638d7e-708b-..."], "groups": ["e3638d7e-708b-..."]}},
  {"update":  {"users": ["e3638d7e-708b-..."], "groups": ["e3638d7e-708b-..."]}},
  {"delete":  {"users": ["e3638d7e-708b-..."], "groups": ["e3638d7e-708b-..."]}}
]
```
Both fixes confirmed working:
- `to_json()` produces plain dicts that identity extraction parses without error
- Pinned permissions collection survives through to file data construction (was `None` without the pin)

File: unstructured_ingest/processes/connectors/sharepoint.py [hunk 3]
```diff
@@ -98,6 +125,146 @@ class SharepointIndexer(OnedriveIndexer):
...
+    @staticmethod
+    def _pin_permissions(drive_item: DriveItem) -> Any:
+        """Pin the permissions collection into properties so it survives across accesses.
+
+        The property getter creates a fresh empty PermissionCollection each time
+        when the key is absent. Pinning ensures batch/query results are retained.
+        """
+        perms = drive_item.permissions
+        drive_item.properties["permissions"] = perms
+        return perms
```
[the remark this answers is not in the document]
Comment (brndnblck) [reply unstructured_ingest/processes/connectors/sharepoint.py:224]: False positive. This has been addressed. Both validation and the regenerated fixtures would show `null` if `execute_batch` was still clearing the pin.

https://github.com/Unstructured-IO/unstructured-ingest/pull/699#issuecomment-4340239945

File: unstructured_ingest/processes/connectors/sharepoint.py [hunk 4]
```diff
@@ -98,6 +125,146 @@ class SharepointIndexer(OnedriveIndexer):
...
+        return user_ids, group_ids
+
+    def extract_permissions(
+        self,
+        permissions: list[Permission],
+    ) -> list[dict[str, Any]]:
+        """Normalize Graph API permissions to the standard read/update/delete format."""
+        if not permissions:
+            logger.debug("no permissions found")
+            return [{}]
```
[the remark this answers is not in the document]
Comment (brndnblck) [reply unstructured_ingest/processes/connectors/sharepoint.py:179]: Won't fix. 

I don't like the inconsistency here either, but this is following a pre-existing convention from the other connectors and is a non-issue in practice since `drive_item_to_file_data_sync` is guarded by `if perms:`.

I chose to stay consistent with existing behavior here, and if we want to revisit that (breaking change on multiple connectors) we should make that change separately and version accordingly.

**Review by potter-potter — APPROVED** [review]
looks good.
